flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateSerializer.java

```diff
@@ -0,0 +1,199 @@
+ */
+
+package org.apache.flink.runtime.checkpoint.channel;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.runtime.io.network.buffer.Buffer;
+import org.apache.flink.runtime.io.network.buffer.BufferBuilder;
+import org.apache.flink.util.Preconditions;
+
+import org.apache.flink.shaded.netty4.io.netty.buffer.ByteBuf;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+
+import static java.lang.Math.addExact;
+import static java.lang.Math.min;
+
+interface ChannelStateSerializer {
+
+	void writeHeader(DataOutputStream dataStream) throws IOException;
+
+	void writeData(DataOutputStream stream, Buffer... flinkBuffers) throws IOException;
+}
+
+interface ChannelStateDeserializer {
+
+	void readHeader(InputStream stream) throws IOException;
+
+	/**
+	 * Reads the length of state.
+	 *
+	 * @return size of state in bytes and offset of the current position resulted from reading.
+	 */
+	Tuple2<Integer, Integer> readLength(InputStream stream) throws IOException;
+
+	int readData(InputStream stream, ChannelStateByteBuffer buffer, int bytes) throws IOException;
+}
+
+/**
+ * Wrapper around various buffers to receive channel state data.
+ */
+@Internal
+@NotThreadSafe
+interface ChannelStateByteBuffer {
+
+	boolean isWritable();
+
+	int writableBytes();
+
+	int writeBytes(InputStream input, int bytesToRead) throws IOException;
+
+	static ChannelStateByteBuffer wrap(Buffer buffer) {
+		return new ChannelStateByteBuffer() {
+
+			private ByteBuf byteBuf = buffer.asByteBuf();
+
+			@Override
+			public boolean isWritable() {
+				return byteBuf.isWritable();
+			}
+
+			@Override
+			public int writableBytes() {
+				return byteBuf.writableBytes();
+			}
+
+			@Override
+			public int writeBytes(InputStream input, int bytesToRead) throws IOException {
+				return byteBuf.writeBytes(input, Math.min(bytesToRead, byteBuf.writableBytes()));
+			}
+		};
+	}
+
+	static ChannelStateByteBuffer wrap(BufferBuilder bufferBuilder) {
+		final byte[] buf = new byte[1024];
+		return new ChannelStateByteBuffer() {
+			@Override
+			public boolean isWritable() {
+				return !bufferBuilder.isFull();
+			}
+
+			@Override
+			public int writableBytes() {
+				return bufferBuilder.writableBytes();
+			}
+
+			@Override
+			public int writeBytes(InputStream input, int upToBytes) throws IOException {
+				int left = upToBytes;
+				for (int toRead = getToRead(left); toRead > 0; toRead = getToRead(left)) {
+					int read = input.read(buf, 0, toRead);
+					int copied = bufferBuilder.append(java.nio.ByteBuffer.wrap(buf, 0, read));
+					Preconditions.checkState(copied == read);
+					left -= read;
+				}
+				bufferBuilder.commit();
+				return upToBytes - left;
+			}
+
+			private int getToRead(int bytesToRead) {
+				return min(bytesToRead, min(buf.length, writableBytes()));
```

Comment:
Use `bufferBuilder.writableBytes()` to replace `writableBytes()` then we can remove this explicit interface method.